exercises/matrix/src/test/java/MatrixTest.java

```diff
@@ -0,0 +1,208 @@
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.experimental.runners.Enclosed;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(Enclosed.class)
+public class MatrixTest {
+
+    @RunWith(Parameterized.class)
+    public static class GetFirstRowTest {
+        private Matrix matrix;
+        private int[] firstRow;
+
+        @Parameterized.Parameters(name = "{index}: checking first row of matrix constructed with string \n\"{0}\".")
+        public static Collection<Object[]> data() {
+            return Arrays.asList(new Object[][]{
+                    {"0", new int[] {0}},
+                    {"0 1", new int[] {0, 1}},
+                    {  "0\n"
+                     + "1", new int[] {0}},
+                    {  "0 1\n"
+                     + "2 3", new int[] {0, 1}},
+                    {  "0 1 2\n"
+                     + "3 4 5\n"
+                     + "6 7 8", new int[] {0, 1, 2}}
+            });
+        }
+
+        public GetFirstRowTest(String matrixAsString, int[] firstRow) {
+            this.matrix = new Matrix(matrixAsString);
+            this.firstRow = firstRow;
+        }
+
+        @Test
+        public void getFirstRowTest() {
+            assertArrayEquals(firstRow, matrix.getRow(0));
+        }
+    }
+
+    @Ignore
+    @RunWith(Parameterized.class)
+    public static class GetLastRowTest {
+        private Matrix matrix;
+        private int[] lastRow;
+
+        @Parameterized.Parameters(name = "{index}: checking last row of matrix constructed with string \n\"{0}\".")
+        public static Collection<Object[]> data() {
+            return Arrays.asList(new Object[][]{
+                    {"0", new int[] {0}},
+                    {"0 1", new int[] {0, 1}},
+                    {  "0\n"
+                     + "1", new int[] {1}},
+                    {  "0 1\n"
+                     + "2 3", new int[] {2, 3}},
+                    {  "0 1 2\n"
+                     + "3 4 5\n"
+                     + "6 7 8", new int[] {6, 7, 8}}
+            });
+        }
+
+        public GetLastRowTest(String matrixAsString, int[] lastRow) {
+            this.matrix = new Matrix(matrixAsString);
+            this.lastRow = lastRow;
+        }
+
+        @Test
+        public void getLastRowTest() {
+            assertArrayEquals(lastRow, matrix.getRow(matrix.getRowsCount() - 1));
```

Comment:
Since this test uses `getRowsCount` to identify the final row of the matrix, I wonder if we should move the tests for `CountColumnsTest` and `CountRowsTest` classes to the very beginning of the test suite? Otherwise those test classes are effectively redundant, as the implementor will already have had to build working implementations of `getRowsCount` and `getColumnsCount` to pass earlier tests.